python/ray/data/_internal/execution/operators/output_splitter.py

```diff
@@ -0,0 +1,224 @@
+import math
+from typing import List
+
+from ray.data.block import Block, BlockMetadata, BlockAccessor
+from ray.data._internal.remote_fn import cached_remote_fn
+from ray.data._internal.stats import StatsDict
+from ray.data._internal.execution.interfaces import (
+    RefBundle,
+    PhysicalOperator,
+    ExecutionResources,
+)
+from ray.types import ObjectRef
+
+
+class OutputSplitter(PhysicalOperator):
+    """An operator that splits the given data into `n` output splits.
+
+    The output bundles of this operator will have a `bundle.output_split` attribute
+    set to an integer from [0..n-1]. This operator tries to divide the rows evenly
+    across output splits.
```

Comment:
I think we should say something about what ordering guarantee is (not) provided.

Reply:
Note about the status quo: equalized splitting AND/OR locality-aware splitting does _not_ currently preserve order, while `equal=False`, `locality_hints=None` splitting currently does preserve order. These semantics are not documented, but it seems like we could respect the configured order preservation without much difficulty?

Reply:
Hmm what does ordering mean though? We'd want to provide deterministic splitting, which would definitely not be locality aware.

Stepping back, we should probably revisit the deterministic execution later on more holistically.

Reply:
We should probably preserve the existing semantics (equal=False and no locality hints) for backward compability?

Reply:
Yes, we can support these in the new API (this will only be used for a new streaming_split() Dataset API).